Extend service invocation to non-Dapr endpoints docs (#3384)
dapr/docs: daprdocs/content/en/reference/api/service_invocation_api.md

weight: 100
---

Dapr provides users with the ability to call other applications that are using Dapr with a unique named identifier (appId), or HTTP endpoints that are not using Dapr.
This allows applications to interact with one another via named identifiers and puts the burden of service discovery on the Dapr runtime.

## Invoke a method on a remote Dapr app

This endpoint lets you invoke a method in another Dapr enabled app.

### HTTP Request

```
PATCH/POST/GET/PUT/DELETE http://localhost:<daprPort>/v1.0/invoke/<appID>/method/<method-name>
```

## Invoke a method on a non-Dapr endpoint

> This endpoint lets you invoke a method on a non-Dapr endpoint using a HTTPEndpoint resource name, or a Fully Qualified Domain Name (FQDN) URL.

Reviewer: ```suggestion
This endpoint lets you invoke a method on a non-Dapr endpoint using an `HTTPEndpoint` resource name, or a Fully Qualified Domain Name (FQDN) URL.
```

weight: 100
---

> Dapr provides users with the ability to call other applications that have unique ids, or non-Daprized endpoints.

Reviewer 2: ```suggestion
Dapr provides users with the ability to call other applications that are using Dapr with a unique named identifier (appId), or HTTP endpoints that are not using Dapr.
```

> This functionality allows apps to interact with one another via named identifiers and puts the burden of service discovery on the Dapr runtime.

Reviewer 2: ```suggestion
This allows applications to interact with one another via named identifiers and puts the burden of service discovery on the Dapr runtime.
```

> ## Invoke a method on a remote dapr app

Reviewer 2: ```suggestion
## Invoke a method on a remote Dapr app
```

### HTTP Request

```
PATCH/POST/GET/PUT/DELETE http://localhost:<daprPort>/v1.0/invoke/<appID>/method/<method-name>
```

> ## Invoke a method on a non-Daprized endpoint

Reviewer 2: ```suggestion
## Invoke a method on a non-Dapr endpoint
```

This endpoint lets you invoke a method on a non-Daprized endpoint.

### HTTP Request

```

> PATCH/POST/GET/PUT/DELETE http://localhost:<daprPort>/v1.0/invoke/<HTTPEndpoint name or URL>/method/<method-name>

Reviewer 2: ```suggestion
PATCH/POST/GET/PUT/DELETE http://localhost:<daprPort>/v1.0/invoke/<HTTPEndpoint name>/method/<method-name>

PATCH/POST/GET/PUT/DELETE http://localhost:<daprPort>/v1.0/invoke/<FQDN URL>/method/<method-name>
```

Reviewer 2: You also need to 
1) Update the URL Parameters table indicate that there is an order or precedence
2) Are there new error codes returned if the HTTPEndpoint resource or the URL called failed? I would expect to see these and this may need a code change
3) Add curl examples in the examples section of calling the API in both ways.

Author: for number 2, it's the same error responses that you receive with regular service to service invocation. In the case that someone put an invalid base URL or invalid header, then you'd see the usual service invocation error responses indicating the reason for failure.

--------- | -----------
daprPort | the Dapr port

> appId | the App ID associated with the remote app

Reviewer 2: ```suggestion
appID | the App ID associated with the remote app
```

HTTPEndpoint name | the HTTPEndpoint resource associated with the external endpoint

> FQDN URL | Fully Qualified Domain Name url to invoke on the external endpoint

Reviewer 2: ```suggestion
FQDN URL | Fully Qualified Domain Name URL to invoke on the external endpoint
```

### Request received by invoked service

> Once your service code invokes a method in another Dapr enabled app or non-Daprized endpoint, Dapr will send the request, along with the headers and body, on the `<method-name>` endpoint.

Reviewer 2: ```suggestion
Once your service code invokes a method in another Dapr enabled app or non-Daprized endpoint, Dapr sends the request, along with the headers and body, on the `<method-name>` endpoint.
```

In this URL, `testing` is the namespace that `mathService` is running in.

#### Non-Dapr Endpoint Example

> If the `mathService` service was a non-Dapr application, then it could be invoked using service invocation via an HTTPEndpoint as well as a fully qualified domain name URL.

Reviewer 2: ```suggestion
If the `mathService` service was a non-Dapr application, then it could be invoked using service invocation via an `HTTPEndpoint` as well as a fully qualified domain name (FQDN) URL.
```

> If the `mathService` service was a non-Dapr application, then it could be invoked using service invocation via an `HTTPEndpoint` as well as a fully qualified domain name (FQDN) URL.

Reviewer: ```suggestion
If the `mathService` service was a non-Dapr application, then it could be invoked using service invocation via an `HTTPEndpoint`, as well as a Fully Qualified Domain Name (FQDN) URL.
```
